File: 35.searchInsertPosition/35.py
"""
遍历
"""
from typing import List

class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        i = 0
        while i < len(nums):
            if nums[i] < target:
                i +=1
            else:
                break
        return i

"""
标准二分
"""
class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        high = len(nums) - 1 
        low = 0
        ### 特殊情况
        if target < nums[0]:
            return 0
        if target > nums[high]:
            return high+1
        ###
        ans = 0
        while low <= high:
            mid = (high+low)//2
            if nums[mid] == target:
                ans = mid
                return ans # 可以化简，return mid
            elif nums[mid] < target:
                ans = mid + 1
                low = mid + 1
            else:
                high = mid - 1
        return ans # 可以化简，return low
        #  return mid+1 则会出错
        #   因为 ans 只在elif 中更新
        #   如果改为mid + 1 , 则mid的值可能会在else中被更新，从而mid+1 != ans


if __name__=="__main__":

    test=[1,3,5,6]
    solve=Solution()
    print(solve.searchInsert(test,7))

File: 35.searchInsertPosition/35.cpp
#include <iostream>
#include<vector>
using namespace std;

class Solution{
public:
    int searchInsert(std::vector<int>& nums, int target){
        int index =-1; //最终返回的下标，若找到则为该元素下标，否则为需要插入的位置
        int begin = 0;//搜索区间左端点
        int end = nums.size() -1;//搜索区间右端点
        while(index == -1){ //若index==-1,则说明还未找到正确位置（没找到均有正值），持续二分搜索，
            int mid=(begin+end)/2; //计算中心位置
            if(target == nums[mid]){ //若目标值为中间值，则为找到target
                index = mid;
            }
            else if(target<nums[mid]){
                if(mid==0||target > nums[mid-1]){ //如果是在临介点，则插入的位置为中间点
                    index=mid;
                }
                end = mid-1;//如果target小于中点，更新区间右端点。
            }
            else if (target > nums[mid]){
                if (mid == nums.size()-1||target<nums[mid+1]){
                    index=mid+1;
                }
                begin = mid+1;//如果target大于终点，更新区间左端点
            }
        }
        return index;
    }
};

int main(){
    int test[]={1,3,5,6};
    std:vector<int> nums;
    Solution solve;
    for (int i=0;i<4;i++){
        nums.push_back(test[i]);
    }
    for (int i=0;i<8;i++){
        printf("i= %d index =%d\n",i,solve.searchInsert(nums,i));
    }
    return 0;
}